ENH: linalg: add subspace_angles function.

Eric89GXL opened the pull request:
I sent an email to SciPy-dev about this, but opening a PR in case people already think it's worthwhile.

WarrenWeckesser commented:
Not a review, just a quick comment: always include at least one example in the "Examples" section of the docstring when you create a new function.  We would like the list at https://github.com/scipy/scipy/issues/7168 to get smaller, not bigger.

Eric89GXL commented:
Ahh yes of course, I should have known better...

ilayn commented:
I would also kindly ask if the name can be something else. I know this is matlab convention (which is not famous for its consistency) but imho it is a terribly generic and uninformative name. To start with, the command is called subspace but the output is an angle and the arguments are matrices. It is quite analogous to calling the determinant function as `volume(A)`

Moreover the output is in radians but that's a whole other story.

Eric89GXL commented:
> I would also kindly ask if the name can be something else.

I think the best name is probably `subspace_angles`. Typically you'd use this function to get the subspace angle between two matrices, so it's very readable as `subspace_angles(A, B)`.

> Moreover the output is in radians but that's a whole other story.

AFAIK most functions in the scientific ecosystem use radians so it seems like the reasonable choice here. `np.rad2deg` is easy enough to use.

ilayn commented:
I might be too fond of User-eXperience and "if it's not complicated, it should do it" type of frontend design mantra. I think a `angles_in=<rad, deg>` keyword won't hurt. ;) But I wouldn't argue too much about it. Just a minute nudge :)

Eric89GXL commented:
I'd rather not. I don't think we usually have it, so it would make for
inconsistent behavior (and I wouldn't want to add such arguments to all
functions that deal with angles).

ilayn commented:
Got it. No problem.

Eric89GXL commented:
For @WarrenWeckesser:

https://2570-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.linalg.subspace_angles.html#scipy.linalg.subspace_angles

This is ready for review/merge from my end.

Eric89GXL commented:
@ilayn are you happy with this?

Any other `linalg` reviewer have time to look (@pv @rgommers)?

perimosocordiae commented on scipy/linalg/decomp_svd.py:
```diff
@@ -238,3 +238,101 @@ def orth(A):
+
+    # 5. Compute the principal angles
+    theta = empty(len(sigma))
+    for ii, (s, m) in enumerate(zip(sigma, mu)):
```
This loop looks vectorizable.

ilayn replied:
Just to give an example:

    theta = empty_like(sigma)
    ind = sigma ** 2 < 0.5
    theta[ind] = arccos(sigma[ind])
    theta[~ind] = arcsin(mu[~ind])

perimosocordiae commented on scipy/linalg/decomp_svd.py:
```diff
@@ -238,3 +238,101 @@ def orth(A):
+    mu = svdvals(B, overwrite_a=True)
+
+    # 5. Compute the principal angles
+    theta = empty(len(sigma))
```
`theta = np.empty_like(sigma)`

perimosocordiae commented on scipy/linalg/decomp_svd.py:
```diff
@@ -238,3 +238,101 @@ def orth(A):
+    QB = orth(B)
+
+    # 2. Compute SVD for cosine
+    sigma = svdvals(dot(QA.T, QB), overwrite_a=True)
```
Is it worth keeping the result of `dot(QA.T, QB)` around? Or is this cheap enough to (sometimes) recalculate in step 3 without issue?

ilayn replied:
Indeed in the `else` branch of Step 3, it is the transpose of this product that shows up. So either way it is reused.

perimosocordiae commented on scipy/linalg/decomp_svd.py:
```diff
@@ -238,3 +238,101 @@ def orth(A):
+        B = QA - dot(QB, dot(QB.T, QA))
+
+    # 4. Compute SVD for sine
+    mu = svdvals(B, overwrite_a=True)
```
It may be worth deferring this SVD until we know we need it (i.e., if `((sigma**2) >= 0.5).any()`).

perimosocordiae commented on scipy/linalg/tests/test_decomp.py:
```diff
@@ -2502,5 +2503,39 @@ def test_orth():
+    for x in (A, B):
+        assert_allclose(subspace_angles(x, x), np.zeros(x.shape[1]),
+                        atol=1e-14)
+    # From MATLAB
```
Include the matlab expression here for reproducibility?

Eric89GXL replied:
It's just `subspace` called with the two matrices so probably not necessary to give the exact call (it depends on boilerplate loading data), but I should at least mention `subspace` since the name now diverges

ilayn commented on scipy/linalg/decomp_svd.py:
```diff
@@ -238,3 +238,101 @@ def orth(A):
+        raise ValueError('expected matrix')
+    if len(B) != len(A):
+        raise ValueError('A and B must have the same number of rows')
+
```
I think a more informative check can be given here to help the user what the problem is. I've also noticed that when the error mentions a matrix, the users tend to confuse it with `np.matrix`. I can propose a check along the lines below. Since we also don't know what to do with `ndim>2` it might be a better idea to constrain it to exactly to 2D arrays.

    for ind, x in enumerate((A,B)):
        if not x.ndim == 2:
            raise ValueError('Array {} must be a 2D array'.format('AB'[ind]))
    if B.shape[0] != A.shape[0]:
        raise ValueError('A and B must have the same number of rows')

Eric89GXL commented:
@ilayn comments addressed, feel free to look again when you have a chance

ilayn commented:
That looks pretty good to me. Thank you.

Eric89GXL commented:
Added text to release notes, squashed, and CIs are happy. Ready for merge from my end.

CircleCI has rendered the [doc](https://3135-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/generated/scipy.linalg.subspace_angles.html#scipy.linalg.subspace_angles) and [notes](https://3135-1460385-gh.circle-artifacts.com/0/home/ubuntu/scipy/doc/build/html-scipyorg/release.1.0.0.html).

pv commented:
needs rebase (due to tests)

Eric89GXL commented:
@pv fixed, thanks for the ping
